File: apps/api-rs/src/document/docx.rs
use super::shared::*;
use super::*;
pub(super) fn parse_docx(
    doc_id: Uuid,
    parse_job_id: Uuid,
    title: &str,
    bytes: &[u8],
) -> Result<ParsedDocument> {
    let mut archive = open_zip(bytes)?;
    let xml = read_zip_text(&mut archive, "word/document.xml")?;
    let style_levels = read_zip_text(&mut archive, "word/styles.xml")
        .ok()
        .map(|styles| {
            validate_xml_nesting(&styles)?;
            docx_style_heading_levels(&styles)
        })
        .transpose()?
        .unwrap_or_default();
    validate_xml_nesting(&xml)?;
    let doc = roxmltree::Document::parse(&xml).context("invalid_docx_xml")?;
    let body = doc
        .descendants()
        .find(|n| n.tag_name().name() == "body")
        .ok_or_else(|| anyhow!("docx_body_missing"))?;

    let mut blocks = Vec::new();
    let mut tables = Vec::new();
    let mut anchors = Vec::new();
    let mut heading_path: Vec<(i32, String)> = Vec::new();

    let content_nodes = body
        .descendants()
        .filter(|node| {
            if !node.is_element() || !matches!(node.tag_name().name(), "p" | "tbl") {
                return false;
            }
            !node
                .ancestors()
                .skip(1)
                .take_while(|ancestor| *ancestor != body)
                .any(|ancestor| {
                    ancestor.is_element() && matches!(ancestor.tag_name().name(), "p" | "tbl")
                })
        })
        .collect::<Vec<_>>();

    for child in content_nodes {
        match child.tag_name().name() {
            "p" => {
                let text = collect_text(child).trim().to_string();
                if text.is_empty() {
                    continue;
                }
                let style = paragraph_style(child);
                let has_numbering = child.descendants().any(|n| n.tag_name().name() == "numPr");
                let heading_level = paragraph_outline_level(child)
                    .or_else(|| {
                        style
                            .as_ref()
                            .and_then(|style| style_levels.get(style).copied())
                    })
                    .or_else(|| heading_level_from_style(style.as_deref()));
                let block_type = if heading_level.is_some() {
                    "heading"
                } else if has_numbering {
                    "list_item"
                } else {
                    "paragraph"
                };
                if let Some(level) = heading_level {
                    heading_path.retain(|(existing, _)| *existing < level);
                    heading_path.push((level, text.clone()));
                }
                let path = if block_type == "heading" {
                    heading_path
                        .iter()
                        .take(heading_path.len().saturating_sub(1))
                        .map(|(_, h)| h.clone())
                        .collect()
                } else {
                    heading_path.iter().map(|(_, h)| h.clone()).collect()
                };
                blocks.push(ParsedBlock {
                    block_id: Uuid::new_v4(),
                    block_index: blocks.len() as i32,
                    block_type: block_type.to_string(),
                    text,
                    heading_level,
                    heading_path: path,
                    page_start: None,
                    page_end: None,
                    slide_index: None,
                    table_id: None,
                    bbox: None,
                    anchor_ids: vec![],
                    source_ref: json!({"format": "docx", "node": "w:p", "index": blocks.len()}),
                    metadata: json!({"style": style}),
                });
            }
            "tbl" => {
                let block_id = Uuid::new_v4();
                let table = parse_xml_table(
                    child,
                    doc_id,
                    parse_job_id,
                    block_id,
                    tables.len() as i32,
                    heading_path.iter().map(|(_, h)| h.clone()).collect(),
                    None,
                    None,
                    None,
                    "docx",
                );
                blocks.push(ParsedBlock {
                    block_id,
                    block_index: blocks.len() as i32,
                    block_type: "table".to_string(),
                    text: table.markdown.clone(),
                    heading_level: None,
                    heading_path: table.heading_path.clone(),
                    page_start: None,
                    page_end: None,
                    slide_index: None,
                    table_id: Some(table.table_id),
                    bbox: None,
                    anchor_ids: vec![],
                    source_ref: json!({"format": "docx", "node": "w:tbl", "index": tables.len()}),
                    metadata: json!({}),
                });
                tables.push(table);
            }
            _ => {}
        }
    }
    attach_table_cell_anchors(
        doc_id,
        parse_job_id,
        "docx",
        &mut blocks,
        &tables,
        &mut anchors,
    );

    Ok(finalize_parsed(
        doc_id,
        parse_job_id,
        "docx",
        title,
        None,
        blocks,
        tables,
        anchors,
    ))
}

fn paragraph_style(p: roxmltree::Node) -> Option<String> {
    p.descendants()
        .find(|n| n.tag_name().name() == "pStyle")
        .and_then(|n| attr(n, "val"))
        .map(str::to_string)
}

fn paragraph_outline_level(p: roxmltree::Node) -> Option<i32> {
    p.descendants()
        .find(|node| node.tag_name().name() == "outlineLvl")
        .and_then(|node| attr_i32(node, "val"))
        .map(|level| (level + 1).clamp(1, 9))
}

fn docx_style_heading_levels(xml: &str) -> Result<std::collections::BTreeMap<String, i32>> {
    let doc = roxmltree::Document::parse(xml).context("invalid_docx_styles_xml")?;
    let mut levels = std::collections::BTreeMap::new();
    for style in doc
        .descendants()
        .filter(|node| node.is_element() && node.tag_name().name() == "style")
    {
        let Some(style_id) = attr(style, "styleId") else {
            continue;
        };
        let level = style
            .descendants()
            .find(|node| node.tag_name().name() == "outlineLvl")
            .and_then(|node| attr_i32(node, "val"))
            .map(|level| (level + 1).clamp(1, 9))
            .or_else(|| {
                style
                    .descendants()
                    .find(|node| node.tag_name().name() == "name")
                    .and_then(|node| attr(node, "val"))
                    .and_then(|name| heading_level_from_style(Some(name)))
            });
        if let Some(level) = level {
            levels.insert(style_id.to_string(), level);
        }
    }
    Ok(levels)
}

fn heading_level_from_style(style: Option<&str>) -> Option<i32> {
    let normalized = style?.to_ascii_lowercase().replace(' ', "");
    for level in 1..=6 {
        if normalized == format!("heading{level}") || normalized == format!("标题{level}") {
            return Some(level);
        }
    }
    None
}

File: apps/api-bun/src/document/docx.ts
// 移植自 apps/api-rs/src/document/docx.rs —— 行为对齐 Rust 原版，错误信息保持一致

// Word：解析 word/document.xml（标题层级 / 列表 / 表格），样式表提供 styleId -> 标题级别映射

import type JSZip from 'jszip';
import { newUuid } from '../infra/uuid.ts';
import type { SourceAnchor } from '../models/source_anchor.ts';
import {
  attachTableCellAnchors,
  collectText,
  finalizeParsed,
  openZip,
  parseXmlTable,
  readZipText,
} from './shared.ts';
import { attr, attrI32, descendants, parseXml, validateXmlNesting, type XmlElement } from './xml.ts';
import { trimRust } from './text_utils.ts';
import type { ParsedBlock, ParsedDocument, ParsedTable } from './types.ts';

export async function parseDocx(
  docId: string,
  parseJobId: string,
  title: string,
  bytes: Uint8Array,
): Promise<ParsedDocument> {
  const archive = await openZip(bytes);
  const xml = await readZipText(archive, 'word/document.xml');
  const stylesXml = await readZipTextOptional(archive, 'word/styles.xml');
  let styleLevels = new Map<string, number>();
  if (stylesXml !== null) {
    validateXmlNesting(stylesXml);
    styleLevels = docxStyleHeadingLevels(stylesXml);
  }
  validateXmlNesting(xml);
  const doc = parseXml(xml);
  const body = findBody(doc);
  if (body === null) {
    throw new Error('docx_body_missing');
  }

  const blocks: ParsedBlock[] = [];
  const tables: ParsedTable[] = [];
  const anchors: SourceAnchor[] = [];
  let headingPath: Array<[number, string]> = [];

  const contentNodes = descendants(body)
    .filter((ref): ref is { kind: 'element'; element: XmlElement } => ref.kind === 'element')
    .filter((ref) => ref.element.name === 'p' || ref.element.name === 'tbl')
    .filter((ref) => !hasBlockAncestor(ref.element, body))
    .map((ref) => ref.element);

  for (const child of contentNodes) {
    if (child.name === 'p') {
      const text = trimRust(collectText(child));
      if (text.length === 0) continue;
      const style = paragraphStyle(child);
      const hasNumbering = descendants(child).some(
        (ref) => ref.kind === 'element' && ref.element.name === 'numPr',
      );
      const headingLevel = paragraphOutlineLevel(child)
        ?? (style !== null ? styleLevels.get(style) : undefined)
        ?? headingLevelFromStyle(style);
      const blockType = headingLevel !== undefined
        ? 'heading'
        : hasNumbering
          ? 'list_item'
          : 'paragraph';
      if (headingLevel !== undefined) {
        headingPath = headingPath.filter(([existing]) => existing < headingLevel);
        headingPath.push([headingLevel, text]);
      }
      const path = blockType === 'heading'
        ? headingPath.slice(0, Math.max(headingPath.length - 1, 0)).map(([, heading]) => heading)
        : headingPath.map(([, heading]) => heading);
      blocks.push({
        block_id: newUuid(),
        block_index: blocks.length,
        block_type: blockType,
        text,
        heading_level: headingLevel ?? null,
        heading_path: path,
        page_start: null,
        page_end: null,
        slide_index: null,
        table_id: null,
        bbox: null,
        anchor_ids: [],
        source_ref: { format: 'docx', node: 'w:p', index: blocks.length },
        metadata: { style: style ?? null },
      });
    } else if (child.name === 'tbl') {
      const blockId = newUuid();
      const table = parseXmlTable(
        child,
        docId,
        parseJobId,
        blockId,
        tables.length,
        headingPath.map(([, heading]) => heading),
        null,
        null,
        null,
        'docx',
      );
      blocks.push({
        block_id: blockId,
        block_index: blocks.length,
        block_type: 'table',
        text: table.markdown,
        heading_level: null,
        heading_path: [...table.heading_path],
        page_start: null,
        page_end: null,
        slide_index: null,
        table_id: table.table_id,
        bbox: null,
        anchor_ids: [],
        source_ref: { format: 'docx', node: 'w:tbl', index: tables.length },
        metadata: {},
      });
      tables.push(table);
    }
  }

  attachTableCellAnchors(docId, parseJobId, 'docx', blocks, tables, anchors);

  return finalizeParsed(docId, parseJobId, 'docx', title, null, blocks, tables, anchors);
}

/** 对齐 Rust read_zip_text(...).ok()：样式表缺失或不可读时忽略（但内容解析错误仍抛出） */
async function readZipTextOptional(archive: JSZip, name: string): Promise<string | null> {
  try {
    return await readZipText(archive, name);
  } catch {
    return null;
  }
}

function findBody(doc: XmlElement): XmlElement | null {
  for (const ref of descendants(doc)) {
    if (ref.kind === 'element' && ref.element.name === 'body') return ref.element;
  }
  return null;
}

/** 对齐 Rust content_nodes 过滤：p/tbl 且最近的 p/tbl 祖先不存在（body 之前） */
function hasBlockAncestor(node: XmlElement, body: XmlElement): boolean {
  let current = node.parent;
  while (current !== null && current !== body) {
    if (current.name === 'p' || current.name === 'tbl') return true;
    current = current.parent;
  }
  return false;
}

function paragraphStyle(paragraph: XmlElement): string | null {
  for (const ref of descendants(paragraph)) {
    if (ref.kind === 'element' && ref.element.name === 'pStyle') {
      return attr(ref.element, 'val') ?? null;
    }
  }
  return null;
}

function paragraphOutlineLevel(paragraph: XmlElement): number | undefined {
  for (const ref of descendants(paragraph)) {
    if (ref.kind === 'element' && ref.element.name === 'outlineLvl') {
      const level = attrI32(ref.element, 'val');
      if (level === undefined) return undefined;
      return Math.min(Math.max(level + 1, 1), 9);
    }
  }
  return undefined;
}

function docxStyleHeadingLevels(xml: string): Map<string, number> {
  let doc: XmlElement;
  try {
    doc = parseXml(xml);
  } catch (error) {
    throw new Error('invalid_docx_styles_xml:' + messageOf(error));
  }
  const levels = new Map<string, number>();
  for (const ref of descendants(doc)) {
    if (ref.kind !== 'element' || ref.element.name !== 'style') continue;
    const style = ref.element;
    const styleId = attr(style, 'styleId');
    if (styleId === undefined) continue;
    let level: number | undefined;
    for (const inner of descendants(style)) {
      if (inner.kind === 'element' && inner.element.name === 'outlineLvl') {
        const raw = attrI32(inner.element, 'val');
        if (raw !== undefined) level = Math.min(Math.max(raw + 1, 1), 9);
        break;
      }
    }
    if (level === undefined) {
      for (const inner of descendants(style)) {
        if (inner.kind === 'element' && inner.element.name === 'name') {
          level = headingLevelFromStyle(attr(inner.element, 'val') ?? null);
          break;
        }
      }
    }
    if (level !== undefined) {
      levels.set(styleId, level);
    }
  }
  return levels;
}

function headingLevelFromStyle(style: string | null): number | undefined {
  if (style === null) return undefined;
  const normalized = asciiLower(style).replace(/ /g, '');
  for (let level = 1; level <= 6; level += 1) {
    if (normalized === `heading${level}` || normalized === `标题${level}`) {
      return level;
    }
  }
  return undefined;
}

/** 对齐 Rust to_ascii_lowercase：只折叠 ASCII 大写 */
function asciiLower(value: string): string {
  return value.replace(/[A-Z]/g, (ch) => ch.toLowerCase());
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
